converter/codec.go

```diff
@@ -62,16 +62,41 @@ type ZlibCodecOptions struct {
 	AlwaysEncode bool
 }
 
-type zlibCodec struct{ options ZlibCodecOptions }
-
 // NewZlibCodec creates a PayloadCodec for use in NewCodecDataConverter
 // to support zlib payload compression.
 //
 // While this serves as a reasonable example of a compression encoder, callers
 // may prefer alternative compression algorithms for lots of small payloads.
-func NewZlibCodec(options ZlibCodecOptions) PayloadCodec { return &zlibCodec{options} }
+func NewZlibCodec(options ZlibCodecOptions) PayloadCodec {
+	return &codec{
+		newWriter: func(w io.Writer) (io.WriteCloser, error) {
+			return zlib.NewWriter(w), nil
+		},
+		newReader:    zlib.NewReader,
+		alwaysEncode: options.AlwaysEncode,
+		encodingType: "binary/zlib",
+	}
+}
+
+// NewCompressionCodec creates a PayloadCodec for use in NewCodecDataConverter
+// This can be used to pass compression encoders for different needs.
+func NewCompressionCodec(newWriter func(io.Writer) (io.WriteCloser, error), newReader func(io.Reader) (io.ReadCloser, error), encodingType string, alwaysEncode bool) PayloadCodec {
```

Comment:
I am not sure there is value in a generic reader/writer encoder since the interface is so simple to implement. The zlib is more of a sample. See https://github.com/temporalio/samples-go/blob/main/snappycompress/data_converter.go for another sample.

Reply:
Ya, samples-go is actually the reason I thought I'd put this together, cause I copied the snappy converter and then made a gzip one. 
That's fair though, not a big deal.  If others seem interested then I'll update to make it accept an Options param instead.